Allow configurable strings to be duplicates

ferki opened the pull request:
This PR is an attempt to fix #25 by allowing quoted strings specified via `allowed_strings` configuration option to be present more than once.

My first idea was to use space or `,` as the separator between allowed strings, but that might be problematic when someone would like to allow a string containing more than one words.

Since we have a Perl::Critic policy at hand, which uses PPI anyway, I decided to reuse that module to detect any quoted tokens coming in via the configuration option. Hope this approach provides good enough balance between all the factors.

Please review and merge, or let me know how to improve it further.

ferki committed:
Add test for allowed duplicate strings

ferki committed:
Allow configurable strings to be duplicates

gugod commented on lib/Perl/Critic/Policy/TooMuchCode/ProhibitDuplicateLiteral.pm:
```diff
@@ -12,8 +14,21 @@ sub violates {
+
+    my $allowed_strings_config = $self->__get_config->get('allowed_strings');
+
+    if (defined $allowed_strings_config) {
+        my $allowed_strings_parser = PPI::Document->new(\$allowed_strings_config);
+
+        for my $quoted_token (@{$allowed_strings_parser->find('PPI::Token::Quote') ||[]}) {
+            push @allowed_strings, $quoted_token->string;
+        }
+    }
```
parsing "a list of strings" like this seems a bit unconventioal to me, and that would means we need to write some documentation just to describe the basic syntax of this list. I wonder if there is anything to steal from the `Perl::Critic` tframework...

I guess there is probably some way to declare that we want "a list of string literals that could contain whitespaces"  inside `supported_parameters`. Similar to what's been done here:

https://github.com/gugod/Perl-Critic-TooMuchCode/blob/master/lib/Perl/Critic/Policy/TooMuchCode/ProhibitExtraStricture.pm#L12

ferki replied:
> parsing "a list of strings" like this seems a bit unconventioal to me

I agree, though I did not find any other good or pre-existing way to handle a list of "strings-with-spaces" :disappointed: 

Therefore I ended up with the current quoted strings approach as a first approximation. I find quoted strings to be the closest representation of what I expect to end up as list items, and I hope this would lead to the least amount of surprises.

gugod commented:
Hmmm.... I renamed `master`  and unintentionally made this PR non-mergable. I'll merged this PR but I would also like to modify the way it takes parameters. See: https://github.com/gugod/Perl-Critic-TooMuchCode/pull/26#discussion_r698494872

gugod commented:
Merged: https://github.com/gugod/Perl-Critic-TooMuchCode/commit/6464e52f11cb0ccae3a16c814d2a8dbbf3fc6a5c

gugod commented:
https://metacpan.org/dist/Perl-Critic/view/lib/Perl/Critic/DEVELOPER.pod#%22string-list%22

It looks ilke parameters declared as `string list` is always splitted int words and probably not suitible. I guess we have to have our convention.

ferki commented:
> Hmmm.... I renamed master and unintentionally made this PR non-mergable.

Oh, I did not notice the rename...I updated my local clone, and next time I'll open PRs against the `fun` branch!

It's too late now, but it might be useful some other time: [changing the base branch of a pull request](https://docs.github.com/en/github/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-base-branch-of-a-pull-request) is possible.

> Merged: 6464e52

Thank you very much :+1:

gugod commented:
>>   Hmmm.... I renamed master and unintentionally made this PR non-mergable.
> Oh, I did not notice the rename...I updated my local clone, and next time I'll open PRs against the fun branch!

No I don't think that's realted to how you did it... I change the default branch without realizing that it would make existing pull requests impossible to be merged (at least not via github UI).

Thanks for your contribution :-)

BTW since it'll be be easier to use to have just one whitelist parameter, eventually I would like to make it happen and deprecate the `whitelist_numbers` parameters.

I would make one parameter just `whitelist` and make it accept numbers and quoted strings. In att least the next release both `whitelist` and `whitelist_numbers` will be supported, but in new feature `whitelist_numbers` will be removed.
